posts/go-cli-from-scratch.md

```diff
@@ -0,0 +1,768 @@
+	fmt.Println(usage)
+}
+```
+
+If we test this app again it will slowly start feeling like a real CLI.
+
+```shell
+$ go run main.go
+```
+
+We need to add one more thing before moving one. One important aspect of any CLI tool is exit/status codes. Any CLI needs to return **0** when successful and anything greater when an error occurred.
+
+We'll create this `usageAndExit` method prints the usage message and an optional status message.
+
+```go
+// main.go
+func usageAndExit(msg string) {
+	if msg != "" {
+		fmt.Fprint(os.Stderr, msg)
+		fmt.Fprintf(os.Stderr, "\n")
+	}
+
+	flag.Usage()
+	os.Exit(0)
+}
+```
+
+If we update our main method as follows, then test:
+
+```go
+// main.go
+func main() {
+	usageAndExit("")
+}
+```
+
+You should get this cryptic message after running:
+
+```shell
+$ go run main.go
+Usage of /var/folders/47/jlb0vvyx0px72mykl96x41xm0000gp/T/go-build235122164/b001/exe/main:
+```
+
+Not to worry this just means we are trying to access `flag.Usage()` before setting a usage message. If we update the main method again:
+
+```go
+// main.go
+func main() {
+	flag.Usage = func() {
+		fmt.Fprint(os.Stderr, fmt.Sprint(usage))
+	}
+
+	usageAndExit("")
+}
+```
+
+Now it will behave as expected returning the usage message and the correct exit code.
+
+You can confirm by running:
+
+```shell
+$ echo $?
+0
+```
+
+Now that we defined our commands we work on actually creating them.
+
+### Creating the Command struct
+
+First thing we want to define is the base Command Struct.  Each command needs to be able to have its own set of flags and a function defining what to do with those flags. We'll define this struct in a new file `commands.go`
+
+```go
+// command.go
+type Command struct {
+	flags   *flag.FlagSet
+	Execute func(cmd *Command, args []string)
+}
+```
+
+The `flag.Flagset` is just what it sounds like its a collection of flags, these will be defined for each individual subcommand. And Execute is an anonymous function that accepts a Command and an array of arguments. Note, Go treats functions as first-class citizens so we can pass functions as parameters or when defining an instance of a struct. Next, we need a few functions on the Command struct, it needs to:
+
+* Parse and assign flags to its flagset
+* Tell if flags have been parsed
+* Execute a given function
+
+What those turn into is:
+
+1.  Init()
+2.  Called()
+3.  Run()
+
+So the full version will look like this:
+
+```go
+// command.go
+package command
+
+import (
+	"flag"
+)
+
+type Command struct {
+	flags   *flag.FlagSet
+	Execute func(cmd *Command, args []string)
+}
+
+func (c *Command) Init(args []string) error {
+	return c.flags.Parse(args)
+}
+
+func (c *Command) Called() bool {
+	return c.flags.Parsed()
+}
+
+func (c *Command) Run() {
+	c.Execute(c, c.flags.Args())
+}
+```
+
+This way each new subcommand we add will just be of type Command. This will simplify things for the next step as we modify our main method to allow subcommands to be easily added.
```

Comment:
```suggestion
Each new subcommand we add will just be of type Command. This will simplify things for the next step as we modify our main method to allow subcommands to be easily added.
```